### DISCUSSION TITLE
How to invoke SelectAction and CenterAction while loading graph

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#569] [General] [answered]
I am using GLSP's vscode extension(https://github.com/eclipse-glsp/glsp-vscode-integration).

For RequestModelAction from GLSP client, GLSP server will reply SetModelAction after creating/loading GGraph model. If we pass CenterAction/SelectAction(with node IDs) along with SetModelAction then it has no effect on the graph/diagram.

My requirement is to highlight particular node in the graph after loading. If we create our own GGraph model then graph will be loaded at left top corner. So, I decided include Center Action along with SetModelAction as a reply for RequestModelAction. But CenterAction/SelectAction has no effect on the graph.

Any alternate solution for these issues? 

When we reply SetModelAction to GLSP client, graph loading operation will be running. Any way to get feed back from server upon loading so that we can invoke CenterAction/SelectAction  ?

**Comment by tortmayr** [comment, marked as answer]
Hi @sathya-1994,
in the workflow you described the `SetModelAction` has no effect because it is handled before the model updates is completed on the client side (i.e. it tries to select the node before its actually rendered in the diagram).
To handle such use cases we have two different mechanisms in place:

On the client side you can queue actions until the initial diagram model is loaded using the `onceModelInitialized()` promise of the [GLSPActionDispatcher](https://github.com/eclipse-glsp/glsp-client/blob/8c8ed231addf2bee078e0fc65cfa5632ea08e8d3/packages/client/src/base/action-dispatcher.ts#L38). An example usage can e.g. found in the [GLSPDiagramWidget](https://github.com/eclipse-glsp/glsp-theia-integration/blob/3655cc2dcb7836fcb191e96784bcea645b56664e/packages/theia-integration/src/browser/diagram/glsp-diagram-widget.ts#L277) class.

On the server side you can use the `dispatchAfterNextUpdate` method of the `ActionDispatcher` to queue an action and it gets dispatched after the next model update has been completed. It sounds to me like this would be the approach you should use.
In fact we already do something similar in the [CreateNodeOperationHandler](https://github.com/eclipse-glsp/glsp-server/blob/59a57afcb8670e921f616b69db973766fa53f355/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/operations/gmodel/CreateNodeOperationHandler.java#L62) where we queue an action to select the newly created node.
